imap_processing/ultra/l1b/ultra_l1b_extended.py

```diff
@@ -0,0 +1,716 @@
+    index_array = np.arange(len(start_type))
+    index_left = index_array[start_type_left]
+    index_right = index_array[start_type_right]
+
+    yf = np.zeros(len(start_type))
+    d = np.zeros(len(start_type))
+
+    # front position of particle for left shutter (mm)
+    yf_estimate_left = 40.0
+    # front position of particle for right shutter (mm)
+    yf_estimate_right = -40.0
+
+    # Compute adjustments for left start type
+    dy_lut_left = np.round((yf_estimate_left - yb[start_type_left] / 100) * 256 / 81.92)
+    # y adjustment in mm
+    y_adjust_left = get_y_adjust(dy_lut_left) / 100
+    # hundredths of a millimeter
+    yf[index_left] = (yf_estimate_left - y_adjust_left) * 100
+    # distance adjustment in mm
+    distance_adjust_left = np.sqrt(2) * d_slit_foil - y_adjust_left
+    # hundredths of a millimeter
+    d[index_left] = (slit_z - distance_adjust_left) * 100
+
+    # Compute adjustments for right start type
+    dy_lut_right = np.round(
+        (yb[start_type_right] / 100 - yf_estimate_right) * 256 / 81.92
+    )
+    # y adjustment in mm
+    y_adjust_right = get_y_adjust(dy_lut_right) / 100
+    # hundredths of a millimeter
+    yf[index_right] = (yf_estimate_right + y_adjust_right) * 100
+    # distance adjustment in mm
+    distance_adjust_right = np.sqrt(2) * d_slit_foil - y_adjust_right
+    # hundredths of a millimeter
+    d[index_right] = (slit_z - distance_adjust_right) * 100
+
+    return d, yf
+
+
+def get_coincidence_positions(
+    de_dataset: xarray.Dataset, particle_tof: np.ndarray, sensor: str
+) -> tuple[np.ndarray, np.ndarray]:
+    """
+    Calculate coincidence positions.
+
+    Calculate time for electrons to travel back to
+    the coincidence anode (etof) and the x coincidence position (xc).
+
+    The tof measured by the coincidence anode consists of the particle
+    tof from start to stop, plus the time for the electrons to travel
+    back to the coincidence anode.
+
+    Further description is available on pages 34-35 of
+    IMAP-Ultra Flight Software Specification document
+    (7523-9009_Rev_-.pdf).
+
+    Parameters
+    ----------
+    de_dataset : xarray.Dataset
+        Data in xarray format.
+    particle_tof : np.ndarray
+        Particle time of flight (i.e. from start to stop)
+        (tenths of a nanosecond).
+    sensor : str
+        Sensor name.
+
+    Returns
+    -------
+    etof_sorted : np.ndarray
+        Time for the electrons to travel back to
+        coincidence anode (tenths of a nanosecond).
+    xc_sorted : np.ndarray
+        X coincidence position (hundredths of a millimeter).
+    """
+    index_top = np.where(de_dataset["COIN_TYPE"] == 1)[0]
+    index_bottom = np.where(de_dataset["COIN_TYPE"] == 2)[0]
+    index = np.concatenate((index_top, index_bottom))
+
+    # Normalized TDCs
+    # For the stop anode, there are mismatches between the coincidence TDCs,
+    # i.e., CoinN and CoinS. They must be normalized via lookup tables.
+
+    # TpCoinNNorm Top
```

Comment:
I think there is a lot of duplicated code for the two cases here, so would it make sense to try and factor some of this out? I think you could also subset your dataset instead of keeping track of the indices and using those all the itme.
https://docs.xarray.dev/en/v2024.06.0/generated/xarray.Dataset.where.html

```python
top_ds = de_dataset.where(de_dataset["COIN_TYPE"] == 1)
```